### PR TITLE
Fix specifying number of levels with log contour

### PR DESCRIPTION
**Pull request by dstansby**
## PR summary

When plotting contours with a log norm, passing an integer value to the ``levels`` argument to cap the maximum number of contour levels now works as intended.  Partially addresses https://github.com/matplotlib/matplotlib/issues/19856 and replaces https://github.com/matplotlib/matplotlib/pull/25149.

I've maually checked that the added test fails before this fix. Testing with the following script:

```python
import matplotlib.pyplot as plt
from matplotlib.colors import LogNorm
import numpy as np

x, y = np.mgrid[1:10:0.1, 1:10:0.1]
data = np.abs(np.sin(x)*np.exp(y))

for n_levels in range(2, 20):
    fig, ax = plt.subplots()
    im = ax.contourf(x, y, data, norm=LogNorm(), levels=n_levels)
    cbar = fig.colorbar(im, ax=ax)
    n_ticks = len(cbar.ax.get_yticklabels())
    print(f"Requested max {n_levels + 1}, got {n_ticks - 1}")
```

Gives on current `main`:
```
Requested max 3, got 7
Requested max 4, got 7
Requested max 5, got 7
Requested max 6, got 7
Requested max 7, got 7
Requested max 8, got 7
Requested max 9, got 7
Requested max 10, got 7
Requested max 11, got 7
Requested max 12, got 7
Requested max 13, got 7
Requested max 14, got 7
Requested max 15, got 7
Requested max 16, got 7
Requested max 17, got 7
Requested max 18, got 7
Requested max 19, got 7
Requested max 20, got 7
```
and with this PR:

```
Requested max 3, got 3
Requested max 4, got 4
Requested max 5, got 4
Requested max 6, got 4
Requested max 7, got 7
Requested max 8, got 7
Requested max 9, got 7
Requested max 10, got 7
Requested max 11, got 7
Requested max 12, got 7
Requested max 13, got 7
Requested max 14, got 7
Requested max 15, got 7
Requested max 16, got 7
Requested max 17, got 7
Requested max 18, got 7
Requested max 19, got 7
Requested max 20, got 7
```
## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [n/a] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [x] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

File: lib/matplotlib/tests/test_contour.py [hunk 1]
```diff
@@ -221,6 +221,23 @@ def test_log_locator_levels(split_collections):
     _maybe_split_collections(split_collections)
 
 
+@pytest.mark.parametrize("n_levels", [2, 3, 4, 5, 6])
```
Comment (ksunden) [inline lib/matplotlib/tests/test_contour.py:224]: FWIW, only the `n_levels=2` case fails on `main`. 

On main, the `levels` argument doesn't seem to do much of anything, there are always 4 `visible_levels`.

On the PR branch it still is a bit odd, there are only ever 2 (for levels= 2 or 3) or 4 (for levels>=4)

Is this expected?

Comment (dstansby) [reply lib/matplotlib/tests/test_contour.py:224]: I've updated the top comment on this PR with a manual testing script. On `main` the levels are always 7, presumably to avoid too many levels? If requesting `levels=20` and only getting 7 is not desired it seems orthogonal to this fix.

I wonder if the number of levels shown is related to the data being contoured? I've updated the test data to be the same as the example above so the [2, 3, 4] cases fail on main.

File: lib/matplotlib/contour.py [hunk 2]
```diff
@@ -1023,8 +1029,9 @@ def _process_contour_level_args(self, args, z_dtype):
                 levels_arg = 7  # Default, hard-wired.
         else:
             levels_arg = self.levels
+
         if isinstance(levels_arg, Integral):
-            self.levels = self._autolev(levels_arg)
+            self.levels = self._autolev(levels_arg, using_default=self.levels is None)
```
Comment (timhoffm) [inline lib/matplotlib/contour.py:1034]: AFAICS this is the only usage of `_autolev()`. I find the logic of `using_default` a bit hard to follow (in particular also because whether it takes effect depends on `self.logscale`). I suggest to pull the logic in to have `if self.levels is None` instead of `if using_default` inside `_autolev()`.

Comment (dstansby) [reply lib/matplotlib/contour.py]: Thanks for bringing this up - I just stared at the code and got confused myself about the logic. I've refactored slightly, by allowing `N = None` through to `_autolev`, and pushing the hard coding of `N = 7` to `_autolev`. I think this makes more sense now, but it would be good to hear what you think about the latest iteration.

File: lib/matplotlib/contour.py [hunk 3]
```diff
@@ -969,19 +969,29 @@ def _autolev(self, N):
...
+        ----------
+        N : int | None
         """
         if self.locator is None:
             if self.logscale:
-                self.locator = ticker.LogLocator()
+                if N is None:
+                    self.locator = ticker.LogLocator()
+                else:
+                    self.locator = ticker.LogLocator(numticks=N)
```
Comment (timhoffm) [inline lib/matplotlib/contour.py:991]: Actually, `numticks=None` is the default.
```suggestion
                self.locator = ticker.LogLocator(numticks=N)  # N may be None
```

File: lib/matplotlib/contour.py [hunk 4]
```diff
@@ -1011,22 +1021,23 @@ def _process_contour_level_args(self, args, z_dtype):
...
+        if levels_arg is None:
             if args:
+                # Set if levels manually provided
                 levels_arg = args[0]
             elif np.issubdtype(z_dtype, bool):
+                # Set default values for bool data types
                 if self.filled:
                     levels_arg = [0, .5, 1]
                 else:
                     levels_arg = [.5]
```
Comment (timhoffm) [inline lib/matplotlib/contour.py:1034]:
```suggestion
                levels_arg = [0, .5, 1] if self.filled else [.5]
```
looks more concise

File: lib/matplotlib/contour.py [hunk 5]
```diff
@@ -969,19 +969,29 @@ def _autolev(self, N):
...
         We need two more levels for filled contours than for
         line contours, because for the latter we need to specify
         the lower and upper boundary of each range. For example,
         a single contour boundary, say at z = 0, requires only
         one contour line, but two filled regions, and therefore
         three levels to provide boundaries for both regions.
+
+        Parameters
+        ----------
+        N : int | None
```
Comment (timhoffm) [inline lib/matplotlib/contour.py:984]:
```suggestion
        N : int or None
```
to stay in the numpydoc convention.

File: lib/matplotlib/contour.py [hunk 6]
```diff
@@ -969,19 +969,29 @@ def _autolev(self, N):
         Select contour levels to span the data.
 
         The target number of levels, *N*, is used only when the
-        scale is not log and default locator is used.
+        locator is not set and the scale is log or the default
+        locator is used.
```
Comment (timhoffm) [inline lib/matplotlib/contour.py:973]: > If *N* is an int, it is used as the target number of levels.
> Otherwise (N = None), a reasonable default is chosen;
> For logscales the LogLocators default, N=7 otherwise.

**Comment by dstansby** [comment]
Thanks - as well as applying your suggestions, I also factored out the logic to set the locator if it's not already set. This allows the long docstrings to be split up, and makes it explicit in `_process_contour_level_args` where the default locator is being set, instead of hiding it in `_autolev`.

File: lib/matplotlib/contour.py [hunk 7]
```diff
@@ -1011,22 +1023,21 @@ def _process_contour_level_args(self, args, z_dtype):
         """
         Determine the contour levels and store in self.levels.
         """
-        if self.levels is None:
+        levels_arg = self.levels
+        if levels_arg is None:
             if args:
+                # Set if levels manually provided
                 levels_arg = args[0]
```
Comment (timhoffm) [inline lib/matplotlib/contour.py:1030]: This could still be made clearer / the current code is still convoluted (but possibly better as a followup PR to not make too many changes and complicate review)

`args` is not the original `args` but stripped from leading `[X, Y], Z`. Effectively it can only be empty or contain a single element that contains `levels` passed positionally (which is supported, but a bit cumbersome to write out as a signature because the optional [X, Y]` require to absorb all positional parameters in `*args` and we therefore have to reunite a possible positional level from args with the possible kwarg level.

The current implementation also has the surprising side effect that you can do `contour(Z, levels1, levels=levels2)` and the positional `levels1` is sliently ignored.

Since `args` can only be empty or contain the positional levels, we can push the logic out of the function and unite the positional and kwarg paths of `levels` in the caller `_contour_args` - where it belongs logically.

I think eventually we should also make `levels` kw-only. `contour(Z, 5)` or `contour(X, Y, Z, [-1, 0, 1])` are not very readable anyway.

**Review by timhoffm — APPROVED** [review]

File: lib/matplotlib/contour.py [hunk 8]
```diff
@@ -964,12 +964,30 @@ def changed(self):
...
+        If *N* is an int, it is used as the target number of levels.
+        Otherwise when *N* is None, a reasonable default is chosen;
+        for logscales the LogLocator chooses, N=7 is the default
+        otherwise.
 
-        The target number of levels, *N*, is used only when the
-        scale is not log and default locator is used.
+        Parameters
+        ----------
+        N : int or None
```
Comment (QuLogic) [inline lib/matplotlib/contour.py:978]: This really should go in the description of `N`, no?
```suggestion
        Parameters
        ----------
        N : int or None
            If *N* is an int, it is used as the target number of levels.
            Otherwise when *N* is None, a reasonable default is chosen;
            for log scales the LogLocator chooses, N=7 is the default
            otherwise.
```

**Review by QuLogic — APPROVED** [review]

**Comment by dstansby** [comment]
I've milestoned this as 3.11 because it changes longstanding behaviour (even if it is a bugfix)
